### PR TITLE
Extra protection against codegen pessimisation due to unique barriers

### PR DESCRIPTION
**Pull request by anfelor**
This PR changes the `transl_unique_barrier` function to always return `May_be_pushed_down` when uniqueness is not enabled. This makes it easy to reason about the absence of codegen changes for the next compiler roll. To review this, please look through #3066 and #3230 again and convince yourself that if all occurences of `ubr` are `May_be_pushed_down`, then we will emit exactly the same code as before.

### CODE REVIEW

File: lambda/translmode.ml [hunk 1]
```diff
@@ -46,4 +46,10 @@ let transl_modify_mode locality =
 let transl_unique_barrier barrier =
   match Typedtree.Unique_barrier.resolve barrier with
   | Uniqueness.Const.Aliased -> May_be_pushed_down
-  | Uniqueness.Const.Unique -> Must_stay_here
+  | Uniqueness.Const.Unique ->
```
Comment (goldfirere) [inline lambda/translmode.ml:49]: This is what we agreed when talking. But now looking at the code, I wonder: is it ever possible to have `Unique` with the extension off? That would make the statement about the uniqueness extension even stronger.

Comment (anfelor) [reply lambda/translmode.ml:49]: Yes, we could probably return `assert false` here

**Comment by anfelor** [comment]
As discussed in our meeting, this should be good to merge?

**Comment by goldfirere** [comment]
I think we've since decided that it's best to roll back the changes to code-gen entirely, until we can be sure to avoid miscompilation.

**Comment by anfelor** [comment]
Since we decided not to roll back the changes to code-gen, we should merge this. When we add the new flag, this PR ensures that the unique barriers are not enabled unless the user asked for it.

**Pull request merged**